src/persist/src/indexed/runtime.rs

```diff
@@ -72,7 +74,61 @@ where
     })
 }
 
+/// The result of an asynchronous computation.
+///
+/// Unlike [std::future::Future], the computation will complete even if this is
+/// dropped.
+pub struct Future<T>(oneshot::Receiver<Result<T, Error>>);
+
+impl<T> Future<T> {
+    pub(crate) fn new() -> (FutureHandle<T>, Future<T>) {
+        let (tx, rx) = oneshot::channel();
+        (FutureHandle(tx), Future(rx))
+    }
+
+    fn flatten_err(res: Result<Result<T, Error>, oneshot::error::RecvError>) -> Result<T, Error> {
+        match res {
+            Ok(x) => x,
+            // The sender will only hang up if the runtime is no longer running.
+            Err(oneshot::error::RecvError { .. }) => Err(Error::RuntimeShutdown),
+        }
+    }
+
+    /// Blocks and synchronously receives the result.
+    pub fn recv(self) -> Result<T, Error> {
+        futures::executor::block_on(self.into_future())
+    }
+
+    /// Convert this into a [std::future::Future].
+    ///
+    /// The computation represented by self will complete regardless of whether
+    /// the returned Future is polled to completion, but anything derived from
+    /// the returned [std::future::Future] (map, etc) is subject to the normal
+    /// rules.
+    //
+    // TODO: Is it possible for this to impl [std::future::Future] directly?
+    pub fn into_future(self) -> impl std::future::Future<Output = Result<T, Error>> {
+        self.0.map(Self::flatten_err)
+    }
+}
+
+/// A handle for filling the result of an asynchronous computation.
+pub struct FutureHandle<T>(oneshot::Sender<Result<T, Error>>);
+
+impl<T> FutureHandle<T> {
+    pub(crate) fn fill(self, res: Result<T, Error>) {
+        // Don't care if the receiver hung up.
+        let _ = self.0.send(res);
+    }
+}
+
 /// A receiver for the `Result<T, Error>` response of an asynchronous command.
+//
+// TODO: Does this even get us anything now that we have [Future]? The only
```

Comment:
I'm leaning toward doing this TODO in this PR. Worst case, we can pull it back out of git history.

Reply:
Did this TODO